### PR TITLE
[Feature] Support TTA

### PR DESCRIPTION
**Pull request by zytx121**
https://github.com/open-mmlab/mmengine/pull/611 is required.

> ## Motivation
To support Test-Time-Aug (TTA).

## Modification
* Add `DetTTAModel` inherited from the `BaseTTAModel` in MMEngine.
* Add `--tta` option in `tools/test.py` to enable TTA during the test.

## Use cases
**To use the default TTA (average the classification scores of images w/ and w/o flip)**

```shell
 python tools/test.py configs/retinanet/retinanet_r50_fpn_1x_coco.py checkpoints/resnet50_coco.pth --tta
 ```
 
 **To use your own custom TTA method**
 
 1. Add `tta_model` and `tta_pipeline` in your config file:
 
 ```python
tta_model = dict(
    type='DetTTAModel',
    tta_cfg=dict(nms=dict(type='nms', iou_threshold=0.5), max_per_img=100))
tta_pipeline = [
        dict(type='LoadImageFromFile', file_client_args=dict(backend='disk')),
        dict(type='TestTimeAug',
             transforms=[
                 [dict(type='Resize', scale=(1333, 800), keep_ratio=True),
                  dict(type='Resize', scale=(1333, 600), keep_ratio=True)],
                 [dict(type='RandomFlip', prob=1.),
                  dict(type='RandomFlip', prob=0.)],
                 [dict(type='LoadAnnotations', with_bbox=True)],
                 [dict(type='PackDetInputs',
                       meta_keys=('img_id', 'img_path', 'ori_shape',
                                  'img_shape', 'scale_factor', 'flip',
                                  'flip_direction'))]])
    ]
 ```
 
 2. test with TTA
 
 ```shell
 python tools/test.py configs/retinanet/retinanet_r50_fpn_1x_coco.py checkpoints.pth --tta
 ```

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: configs/_base_/models/retinanet_r50_fpn.py [hunk 1]
```diff
@@ -66,3 +66,24 @@
...
         max_per_img=100))
+
+tta_model = dict(type='SingleStageTestTimeAugModel')
+tta_pipeline = [
+    dict(type='LoadImageFromFile', file_client_args=dict(backend='disk')),
+    dict(
+        type='TestTimeAug',
+        transforms=[[
+            dict(type='Resize', scale=(1333, 800), keep_ratio=True),
+            dict(type='Resize', scale=(1333, 600), keep_ratio=True)
```
Comment (hhaAndroid) [inline configs/_base_/models/retinanet_r50_fpn.py:77]: just keep one

File: mmdet/models/test_time_augs/model_wrapper.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                raise ValueError(
+                    f"Invalid flipping direction '{flip_direction}'")
+        recovered_masks.append(mask[None, :].float())
+
+    merged_masks = torch.cat(recovered_masks, 0).mean(dim=0)
+    return merged_masks
+
+
+@MODELS.register_module()
+class TwoStageTestTimeAugModel(BaseTTAModel):
```
Comment (hhaAndroid) [inline mmdet/models/test_time_augs/model_wrapper.py:77]: Why do you need two TTA Model, I think you should add a detailed note.

File: tools/test.py [hunk 3]
```diff
@@ -83,6 +84,11 @@ def main():
     if args.show or args.show_dir:
         cfg = trigger_visualization_hook(cfg, args)
 
+    if args.tta:
+        cfg['test_dataloader']['dataset']['pipeline'] = cfg['tta_pipeline']
```
[the remark this answers is not in the document]
Comment (RangiLyu) [reply tools/test.py:88]: What if the test dataset is a dataset wrapper?

File: configs/_base_/models/retinanet_r50_fpn.py [hunk 4]
```diff
@@ -66,3 +66,27 @@
         score_thr=0.05,
         nms=dict(type='nms', iou_threshold=0.5),
         max_per_img=100))
+
+tta_model = dict(
```
Comment (hhaAndroid) [inline configs/_base_/models/retinanet_r50_fpn.py:70]: Are the TTA results aligned with mmdet 2.x?

File: mmdet/models/test_time_augs/det_tta.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        return bboxes
+    else:
+        scores = torch.cat(aug_scores, dim=0)
+        return bboxes, scores
+
+
+@MODELS.register_module()
+class DetTTAModel(BaseTTAModel):
+    """Merge augmented detection results, only bboxes corresponding score under
+    flipping and multi-scale resizing can be processed now."""
```
Comment (hhaAndroid) [inline mmdet/models/test_time_augs/det_tta.py:45]: Provide examples

File: mmdet/models/test_time_augs/det_tta.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    def __init__(self, tta_cfg=None, **kwargs):
+        super().__init__(**kwargs)
+        self.tta_cfg = tta_cfg
+
+    def merge_preds(self, data_samples_list: List[List[DetDataSample]]):
+        """Merge predictions of enhanced data to one prediction.
+
+        Args:
+            data_samples_list (List[List[ClsDataSample]]): List of predictions
```
Comment (hhaAndroid) [inline mmdet/models/test_time_augs/det_tta.py:55]: The meaning of list[list] needs to be explained

Comment (hhaAndroid) [inline mmdet/models/test_time_augs/det_tta.py:55]:
```suggestion
            data_samples_list (List[List[DetDataSample]]): List of predictions
```

File: mmdet/models/test_time_augs/det_tta.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+                of all enhanced data.
+        Returns:
+            List[ClsDataSample]: Merged prediction.
```
Comment (hhaAndroid) [inline mmdet/models/test_time_augs/det_tta.py:58]:
```suggestion
            List[DetDataSample]: Merged prediction.
```

File: mmdet/models/test_time_augs/det_tta.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    bboxes = torch.cat(recovered_bboxes, dim=0)
+    if aug_scores is None:
+        return bboxes
+    else:
+        scores = torch.cat(aug_scores, dim=0)
+        return bboxes, scores
+
+
+@MODELS.register_module()
+class DetTTAModel(BaseTTAModel):
```
Comment (hhaAndroid) [inline mmdet/models/test_time_augs/det_tta.py:43]: Does the instance seg task require a new TTAModel?

File: configs/_base_/models/retinanet_r50_fpn.py (hunk 4 above)
Comment (zytx121) [reply configs/_base_/models/retinanet_r50_fpn.py:70]: Yes, the results have aligned with mmdet 2.26.0.

File: mmdet/models/test_time_augs/det_tta.py (hunk 8 above)
Comment (zytx121) [reply mmdet/models/test_time_augs/det_tta.py:15]: This PR only cover the TTA of object detection.

File: configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py [hunk 9]
```diff
@@ -133,3 +133,41 @@
...
+tta_pipeline = [
+    dict(
+        type='LoadImageFromFile',
+        to_float32=True,
+        file_client_args=dict(backend='disk')),
+    dict(
+        type='TestTimeAug',
+        transforms=[
+            [
+                # ``RandomFlip`` must be placed before ``RandomCenterCropPad``
```
Comment (hhaAndroid) [inline configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py:150]: need to explain why.

File: mmdet/models/test_time_augs/det_tta.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+            of enhanced data which come form one image.
+        Returns:
+            List[DetDataSample]: Merged prediction.
+        """
+        aug_bboxes = []
+        aug_scores = []
+        aug_labels = []
+        img_metas = []
+        for data_sample in data_samples:
+            aug_bboxes.append(data_sample.pred_instances.bboxes)
```
Comment (RangiLyu) [inline mmdet/models/test_time_augs/det_tta.py:116]: Will instance segmentation TTA be supported in this PR? If not, raise an error if there are masks in the pred_instances and add TODO here.

File: configs/_base_/models/retinanet_r50_fpn.py [hunk 11]
```diff
@@ -66,3 +66,25 @@
...
```
Comment (RangiLyu) [inline configs/_base_/models/retinanet_r50_fpn.py:70]: So the TTA config will write in the model's config instead of being an independent file? @HAOCHENYE Please take a look.

**Review by RangiLyu — COMMENTED** [review]
Remember to update the document of TTA after the usage is confirmed.

**Comment by hhaAndroid** [comment]
@zytx121 Please synchronize the latest code to ensure that ci can pass

File: configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py [hunk 12]
```diff
@@ -133,3 +133,43 @@
...
+    dict(
+        type='LoadImageFromFile',
+        to_float32=True,
+        file_client_args=dict(backend='disk')),
+    dict(
+        type='TestTimeAug',
+        transforms=[
+            [
+                # ``RandomFlip`` must be placed before ``RandomCenterCropPad``,
+                # otherwise bounding box coordinates after flipping cannot be
```
Comment (hhaAndroid) [inline configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py:151]: I don't understand why? can you explain?

File: tools/test.py [hunk 13]
```diff
@@ -83,6 +86,40 @@ def main():
...
         cfg = trigger_visualization_hook(cfg, args)
 
+    if args.tta:
+        if 'tta_model' not in cfg:
+            cfg.tta_model = dict(
+                type='DetTTAModel',
+                tta_cfg=dict(
+                    nms=dict(type='nms', iou_threshold=0.5), max_per_img=100))
+        if 'tta_pipeline' not in cfg:
+            if cfg.test_dataloader.dataset.type in ('MultiImageMixDataset',
```
Comment (hhaAndroid) [inline tools/test.py:96]: This way of writing is not general enough. You can refer to the previous similar writing. https://github.com/open-mmlab/mmdetection/blob/master/tools/analysis_tools/optimize_anchors.py#L345

File: configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py (hunk 12 above)
Comment (zytx121) [reply configs/centernet/centernet_r18-dcnv2_8xb16-crop512-140e_coco.py:151]: If we want to recover the flipped result after padding, we need to know how many pixels are padding in the upper left corner of the image. This problem can be avoided by using flip first and then padding.

**Comment by zytx121** [comment]
> @zytx121 Please synchronize the latest code to ensure that ci can pass

Done.

File: configs/_base_/tta/centernet_tta.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
+tta_model = dict(
+    type='DetTTAModel',
+    tta_cfg=dict(nms=dict(type='nms', iou_threshold=0.5), max_per_img=100))
+
```
Comment (RangiLyu) [inline configs/_base_/tta/centernet_tta.py:4]: It would be better to place the model-related TTA files in the model's config folder.

**Review by RangiLyu — DISMISSED** [review]
LGTM. It would be better if you can add a warning in test.py if there is no tta_cfg in the config.

**Comment by RangiLyu** [comment]
Add the use case to docs/en/user_guides/test.md

**Comment by zytx121** [comment]
> LGTM. It would be better if you can add a warning in test.py if there is no tta_cfg in the config.

If there is no tta_cfg in the config, we will set a default tta_cfg for it in `test.py`.
https://github.com/open-mmlab/mmdetection/blob/2a509ccdefda5e7965cfa7d0e8a25f67d772b6ef/tools/test.py#L90-L115

**Comment by zytx121** [comment]
> Add the use case to docs/en/user_guides/test.md

> LGTM. It would be better if you can add a warning in test.py if there is no tta_cfg in the config.

Done.

File: docs/en/user_guides/test.md [hunk 15]
```diff
@@ -198,3 +198,106 @@ data = dict(train_dataloader=dict(...), val_dataloader=dict(...), test_dataloade
 ```
 
 Or you can set it through `--cfg-options` as `--cfg-options test_dataloader.batch_size=2`
+
+## Test Time Augmentation (TTA)
+
+Test time augmentation (TTA) is a data augmentation strategy used during the testing phase. It involves applying various augmentations, such as flipping and scaling, to the same image and then merging the predictions of each augmented image to produce a more accurate prediction.  To make it easier for users to use TTA, MMEngine provides [BaseTTAModel](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.model.BaseTTAModel.html#mmengine.model.BaseTTAModel) class, which allows users to implement different TTA strategies by simply extending the BaseTTAModel class according to their needs.
```
Comment (RangiLyu) [inline docs/en/user_guides/test.md:204]:
```suggestion
Test time augmentation (TTA) is a data augmentation strategy used during the testing phase. It involves applying various augmentations, such as flipping and scaling, to the same image and then merging the predictions of each augmented image to produce a more accurate prediction. To make it easier for users to use TTA, MMEngine provides [BaseTTAModel](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.model.BaseTTAModel.html#mmengine.model.BaseTTAModel) class, which allows users to implement different TTA strategies by simply extending the BaseTTAModel class according to their needs.
```

**Review by RangiLyu — APPROVED** [review]

**Review by hhaAndroid — DISMISSED** [review]

File: docs/en/user_guides/test.md [hunk 16]
```diff
@@ -198,3 +198,106 @@ data = dict(train_dataloader=dict(...), val_dataloader=dict(...), test_dataloade
...
+        ], [
+            dict(
+               type='PackDetInputs',
+               meta_keys=('img_id', 'img_path', 'ori_shape',
+                       'img_shape', 'scale_factor', 'flip',
+                       'flip_direction'))
+       ]])]
+```
+
+The above data augmentation pipeline will first perform 3 multi-scaling enhancements on the image, followed by 2 flipping enhancements (flipping and not flipping). Finally, the image is packaged into the final result using PackDetInputs.
```
Comment (ZwwWayne) [inline docs/en/user_guides/test.md:294]: enhancements to transformation, also change the whole doc.

File: docs/en/user_guides/test.md [hunk 17]
```diff
@@ -198,3 +198,106 @@ data = dict(train_dataloader=dict(...), val_dataloader=dict(...), test_dataloade
 ```
 
 Or you can set it through `--cfg-options` as `--cfg-options test_dataloader.batch_size=2`
+
+## Test Time Augmentation (TTA)
+
+Test time augmentation (TTA) is a data augmentation strategy used during the testing phase. It involves applying various augmentations, such as flipping and scaling, to the same image and then merging the predictions of each augmented image to produce a more accurate prediction. To make it easier for users to use TTA, MMEngine provides [BaseTTAModel](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.model.BaseTTAModel.html#mmengine.model.BaseTTAModel) class, which allows users to implement different TTA strategies by simply extending the BaseTTAModel class according to their needs.
```
Comment (ZwwWayne) [inline docs/en/user_guides/test.md:204]:
```suggestion
Test time augmentation (TTA) is a data augmentation strategy used during the test phase. It applies different augmentations, such as flipping and scaling, to the same image for model inference, and then merges the predictions of each augmented image to obtain more accurate predictions. To make it easier for users to use TTA, MMEngine provides [BaseTTAModel](https://mmengine.readthedocs.io/en/latest/api/generated/mmengine.model.BaseTTAModel.html#mmengine.model.BaseTTAModel) class, which allows users to implement different TTA strategies by simply extending the BaseTTAModel class according to their needs.
```

File: docs/en/user_guides/test.md [hunk 18]
```diff
@@ -198,3 +198,106 @@ data = dict(train_dataloader=dict(...), val_dataloader=dict(...), test_dataloade
...
+        ], [
+            dict(
+               type='PackDetInputs',
+               meta_keys=('img_id', 'img_path', 'ori_shape',
+                       'img_shape', 'scale_factor', 'flip',
+                       'flip_direction'))
+       ]])]
+```
+
+Second, you can simplely run:
```
Comment (ZwwWayne) [inline docs/en/user_guides/test.md:239]: Second, set `--tta` when running the test scripts as examples below:

File: configs/centernet/centernet_tta.py [hunk 19]
```diff
@@ -0,0 +1,39 @@
+tta_model = dict(
+    type='DetTTAModel',
```
Comment (ZwwWayne) [inline configs/centernet/centernet_tta.py:2]: Check whether this follows the official implementation of CenterNet. (I thought it is different from the original one, which simply fuse the score of model predictions) If not, we should add comments to indicate that this is different from the TTA of official CenterNet
